File: sdk/java-sdk-protobuf/src/main/java/kalix/javasdk/view/ViewProvider.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.javasdk.view;

import com.google.protobuf.Descriptors;
import kalix.javasdk.impl.MessageCodec;
import kalix.javasdk.impl.view.ViewUpdateRouter;

import java.util.Optional;

public interface ViewProvider {

  Descriptors.ServiceDescriptor serviceDescriptor();

  String viewId();

  ViewOptions options();

  ViewUpdateRouter newRouter(ViewCreationContext context);

  Descriptors.FileDescriptor[] additionalDescriptors();

  default Optional<MessageCodec> alternativeCodec() {
    return Optional.empty();
  }
}

File: sdk/scala-sdk-protobuf/src/main/scala/kalix/scalasdk/view/ViewProvider.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.scalasdk.view

import com.google.protobuf.Descriptors
import scala.collection.immutable.Seq

import kalix.scalasdk.impl.view.ViewUpdateRouter

trait ViewProvider {
  def serviceDescriptor: Descriptors.ServiceDescriptor

  def viewId: String

  def options: ViewOptions

  def newRouter(context: ViewCreationContext): ViewUpdateRouter

  def additionalDescriptors: Seq[Descriptors.FileDescriptor]
}
